core/src/main/java/io/questdb/cutlass/text/CairoTextWriter.java

```diff
@@ -320,14 +325,12 @@ public CharSequence getTableName() {
         }
 
         @Override
-        public int getTimestampIndex() {
-            // not yet on protocol
-            return -1;
-        }
+        public int getTimestampIndex() { return timestampIndex; }
 
         TableStructureAdapter of(ObjList<CharSequence> names, ObjList<TypeAdapter> types) {
             this.names = names;
             this.types = types;
+            timestampIndex = names.indexOf(timestampIndexCol);
```

Comment:
the name & type check can be done here, here is example:

```java
        TableStructureAdapter of(ObjList<CharSequence> names, ObjList<TypeAdapter> types) throws TextException {
            this.names = names;
            this.types = types;
            if (timestampIndexCol == null) {
                timestampIndex = -1;
            } else {
                timestampIndex = names.indexOf(timestampIndexCol);
                if (timestampIndex == -1) {
                    throw TextException.$("invalid timestamp column '").put(timestampIndexCol).put('\'');
                }
                if (types.getQuick(timestampIndex).getType() != ColumnType.TIMESTAMP) {
                    throw TextException.$("not a timestamp '").put(timestampIndexCol).put('\'');
                }
            }
            return this;
        }
```

Reply:
Additionally at end of `prepareTable` we could get timestamp adapter only once:

```java
timestampAdapter = (TimestampAdapter) types.getQuick(timestampIndex);
```

Additionally i would suggest two lambdas returning `Row`. One lambda will use timestamp index, the other will get row without timestamp. This will simplify `onFields`

Reply:
I didn't even notice `prepareTable` existed. As for the lambdas, doing the try catch inside a lambda and then having some control flow in `onFields` seemed more complex than just leaving the check at the start of `onFields`.